### ISSUE TITLE
cannot import js2py in Python 3.6

### ISSUE THREAD

**Issue by leobelen** [PiotrDabkowski/Js2Py#68]
It seems that the library is not compatible with Python 3.6. Importing it results in:

Traceback (most recent call last):
  File "/Applications/Eclipse.app/Contents/Eclipse/plugins/org.python.pydev_5.4.0.201611281236/pysrc/_pydev_bundle/pydev_monkey.py", line 574, in __call__
    return self.original_func(*self.args, **self.kwargs)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/utils/autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/core/management/commands/runserver.py", line 113, in inner_run
    autoreload.raise_last_exception()
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/utils/autoreload.py", line 249, in raise_last_exception
    six.reraise(*_exception)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/utils/autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/__init__.py", line 27, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/apps/registry.py", line 108, in populate
    app_config.import_models(all_models)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/django/apps/config.py", line 199, in import_models
    self.models_module = import_module(models_module_name)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 978, in _gcd_import
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load
  File "<frozen importlib._bootstrap>", line 950, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 205, in _call_with_frames_removed
  File "/docs/projects/www/antares/antares/apps/core/models/hrn_code.py", line 7, in <module>
    import js2py
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/js2py/__init__.py", line 68, in <module>
    from .base import PyJsException
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/js2py/base.py", line 1563, in <module>
    FunctionPrototype = PyJsFunction(Empty, ObjectPrototype)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/js2py/base.py", line 1215, in __init__
    cand = fix_js_args(func)
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/js2py/utils/injector.py", line 23, in fix_js_args
    code = append_arguments(six.get_function_code(func), ('this','arguments'))
  File "/Users/leobelen/.virtualenvs/antares/lib/python3.6/site-packages/js2py/utils/injector.py", line 84, in append_arguments
    raise ValueError("a name was lost in translation")

**Comment by zmwangx** [comment]
Bytecode has been replaced by wordcode in CPython 3.6, among other things; see [CPython bytecode changes](https://docs.python.org/3.6/whatsnew/3.6.html#cpython-bytecode-changes) in *What's New In Python 3.6*. For instance, `js2py.base.Empty.__code__.co_code` used to be `b'\x74\x00\x00\x64\x00\x00\x83\x01\x00\x53'`; now it's `b'\x74\x00\x64\x00\x83\x01\x53\x00'`.

I'm not familiar with bytecode/wordcode myself, but I skimmed [`wordcode.md`](https://github.com/abarnert/cpython/blob/wpy/Python/wordcode.md) and put together this quick patch for the bytecode reader, `js2py.utils.injector.instructions`:

```diff
diff --git a/js2py/utils/injector.py b/js2py/utils/injector.py
index 3ced868..2c724f7 100644
--- a/js2py/utils/injector.py
+++ b/js2py/utils/injector.py
@@ -2,6 +2,7 @@ __all__ = ['fix_js_args']
 
 import types
 import opcode
+import sys
 import six
 
 if six.PY3:
@@ -129,23 +130,38 @@ def append_arguments(code_obj, new_locals):
 
 
 def instructions(code):
-    if six.PY2:
-        code = map(ord, code)
-    i, L = 0, len(code)
-    extended_arg = 0
-    while i < L:
-        op = code[i]
-        i+= 1
-        if op < opcode.HAVE_ARGUMENT:
-            yield [op, None]
-            continue
-        oparg = code[i] + (code[i+1] << 8) + extended_arg
+    if sys.version_info < (3, 6):
+        if six.PY2:
+            code = map(ord, code)
+        i, L = 0, len(code)
         extended_arg = 0
-        i += 2
-        if op == opcode.EXTENDED_ARG:
-            extended_arg = oparg << 16
-            continue
-        yield [op, oparg]
+        while i < L:
+            op = code[i]
+            i+= 1
+            if op < opcode.HAVE_ARGUMENT:
+                yield [op, None]
+                continue
+            oparg = code[i] + (code[i+1] << 8) + extended_arg
+            extended_arg = 0
+            i += 2
+            if op == opcode.EXTENDED_ARG:
+                extended_arg = oparg << 16
+                continue
+            yield [op, oparg]
+    else:
+        i, L = 0, len(code)
+        extended_arg = 0
+        while i < L:
+            op = code[i]
+            oparg = code[i+1] + extended_arg
+            extended_arg = 0
+            i += 2
+            if op < opcode.HAVE_ARGUMENT:
+                yield [op, None]
+            elif op == opcode.EXTENDED_ARG:
+                extended_arg = oparg << 8
+            else:
+                yield [op, oparg]
 
 def write_instruction(inst):
     op, oparg = inst
```

I'm not sure if the endianness of the CPU architecture needs to be taken into account, but this reader should be correct for little-endian architectures like x86 and x64.

I tested this patch, unfortunately it only seems to work for basic cases, like the first few lines in the README example:

```
# Adding Python built-in sum to the JS context:
>>> context = js2py.EvalJs({'python_sum': sum})  
>>> js_code = '''
var a = 10
function f(x) {return x*x}
'''
>>> context.execute(js_code)
# Get value of variable a:
>>> context.a
10
# context.f behaves just like js function so you can supply more than 1 argument. '9'*'9' in javascript is 81.
>>> context.f('9', 0)  
81    
```

but often fails with `SystemError: unknown opcode`. Apparently more (and probably more involved) work is needed.

**Comment by PiotrDabkowski** [comment]
Fixed: a20f3841c8bde4b218faeb92e7fd28d1ff8af4b1.  Thank you for your help @zmwangx! I used dis.Bytecode to interpret new wordcode - its simpler, more reliable and will support future changes to the wordcode. I guess you had SystemError because you did not modify write_instruction function.

**Comment by zmwangx** [comment]
> you did not modify write_instruction function.

Oh right... Thanks for the quick fix!
